components/script/dom/cssstyledeclaration.rs

```diff
@@ -235,21 +238,88 @@ impl CSSStyleDeclaration {
+            CSSStyleOwner::Element(ref element) => element.upcast::<Node>(),
+        };
+
+        if !node.is_connected() {
+            // TODO: Node should be matched against the style rules of this window.
+            // Firefox is currently the only browser to implement this.
+            return DOMString::new();
+        }
+
+        if let Some(result) = self.try_to_get_resolved_style(node, &property) {
+            return result.into();
+        }
+
+        let addr = node.to_trusted_node_address();
+        window_from_node(node).resolved_style_query(addr, self.pseudo.clone(), property)
+    }
+
+    // Try to get the resolved style for a node without having to ask the layout
+    // thread. This is possible when we can simply return the computed value or
+    // we can calculate the used value without looking at layout data structures.
+    // Some properties resovled values compute are their computed values.
+    // See: https://drafts.csswg.org/cssom/#dom-window-getcomputedstyle
+    fn try_to_get_resolved_style(&self, node: &Node, property: &PropertyId) -> Option<String> {
+        let style = match self.pseudo {
+            Some(ref pseudo) => node.style_for_pseudo(pseudo),
+            None => node.style(),
+        }?;
+
+        let longhand_id = match *property {
+            PropertyId::LonghandAlias(id, _) | PropertyId::Longhand(id) => id,
+            // Firefox returns blank strings for the computed value of shorthands,
+            // so this should be web-compatible.
+            PropertyId::ShorthandAlias(..) | PropertyId::Shorthand(_) => return Some(String::new()),
+            PropertyId::Custom(ref name) => {
+                return Some(style.computed_value_to_string(PropertyDeclarationId::Custom(name)));
             },
         }
+        .to_physical(style.writing_mode);
+
+        let computed_style =
+            || style.computed_value_to_string(PropertyDeclarationId::Longhand(longhand_id));
+
+        // For line height, the resolved value is the computed value if it
+        // is "normal" and the used value otherwise.
+        if longhand_id == LonghandId::LineHeight {
+            let font_size = style.get_font().font_size.size.0;
+            return Some(match style.get_inherited_text().line_height {
+                LineHeight::Normal => computed_style(),
+                LineHeight::Number(value) => (font_size * value.0).to_css_string(),
+                LineHeight::Length(value) => value.0.to_css_string(),
+            });
+        }
+
+        // The properties that we calculate below all resolve to the computed value
+        // when the element is display:none or display:contents.
+        let display = style.get_box().display;
+        if display.is_none() || display.is_contents() {
+            return Some(computed_style());
+        }
+
+        let positioned = style.get_box().position != Position::Static;
+        match longhand_id {
```

Comment:
It's a bit unfortunate to have this list of properties duplicated in multiple places :/